### PR TITLE
Support `logfire.instrument` without arguments

### LINKED ISSUE #537: `@logfire.instrument` should work in bear mode
### Description

e.g.

```py
@logfire.instrument
def foobar(a, b, c):
    ...
```

E.g. not require `@logfire.instrument()`.

### PR DESCRIPTION
**Pull request by Kludex**
- Closes #537

### CODE REVIEW

File: logfire/_internal/main.py [hunk 1]
```diff
@@ -528,14 +540,27 @@ def span(
...
         msg_template: LiteralString | None = None,
         *,
         span_name: str | None = None,
         extract_args: bool | Iterable[str] = True,
         allow_generator: bool = False,
-    ) -> Callable[[Callable[P, R]], Callable[P, R]]:
+    ) -> Callable[[Callable[P, R]], Callable[P, R]]: ...
+
+    @overload
+    def instrument(self, msg_template: Callable[P, R]) -> Callable[P, R]: ...
```
Comment (alexmojaki) [inline logfire/_internal/main.py]:
```suggestion
    def instrument(self, func: Callable[P, R]) -> Callable[P, R]: ...
```

File: logfire/_internal/main.py [hunk 2]
```diff
@@ -528,14 +540,27 @@ def span(
...
+
+    def instrument(
+        self,
+        msg_template: Callable[P, R] | LiteralString | None = None,
```
Comment (alexmojaki) [inline logfire/_internal/main.py:558]:
```suggestion
        msg_template_or_func: Callable[P, R] | LiteralString | None = None,
```

File: logfire/_internal/main.py [hunk 3]
```diff
@@ -557,6 +582,8 @@ def my_function(a: int):
             allow_generator: Set to `True` to prevent a warning when instrumenting a generator function.
                 Read https://logfire.pydantic.dev/docs/guides/advanced/generators/#using-logfireinstrument first.
         """
+        if callable(msg_template):
+            return instrument(self, tuple(self._tags), None, span_name, extract_args, allow_generator)(msg_template)
```
Comment (alexmojaki) [inline logfire/_internal/main.py]:
```suggestion
        if callable(msg_template_or_func):
            return self.instrument()(msg_template_or_func)
```

**Comment by Kludex** [comment]
I didn't do what the suggestions suggest because it's a breaking change... People could be doing `msg_template=...`.

File: logfire/_internal/main.py (hunk 1 above)
Comment (Kludex) [reply logfire/_internal/main.py]: This is not possible. The parameter names must match.

File: logfire/_internal/main.py (hunk 3 above)
Comment (Kludex) [reply logfire/_internal/main.py]: This I can do.
```suggestion
        if callable(msg_template):
            return self.instrument()(msg_template)
```

**Comment by alexmojaki** [comment]
OK leave the main signature as is but use the name `func` in the overload and silence pyright about it.

**Comment by Kludex** [comment]
> OK leave the main signature as is but use the name `func` in the overload and silence pyright about it.

The problem is the typing issue is not on the overload method, but on the main function itself...

Screenshot 2024-11-18 at 10 56 28

We may make typing mistakes in the `instrument` without noticing if we do this.

**Comment by alexmojaki** [comment]
What if you just ignore that specific error code?

**Comment by Kludex** [comment]
> What if you just ignore that specific error code?

I'm not sure if this is working as expected. But I can do it.

**Review by alexmojaki — APPROVED** [review]

**Pull request merged**